src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp

```diff
@@ -0,0 +1,113 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <pup.h>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "Domain/Amr/Flag.hpp"
+#include "Domain/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PowerMonitors.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Options/String.hpp"
+#include "ParallelAlgorithms/Amr/Criteria/Criterion.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+template <size_t>
+class ElementId;
+/// \endcond
+
+namespace amr::Criteria {
+
+/*!
+ * \brief Refine the grid towards the target truncation error
+ *
+ * If any tensor component has a truncation error above the target value, the
+ * element will be p-refined. For details on how the truncation error is
+ * computed see `PowerMonitors::truncation_error`.
+ *
+ * \tparam Dim Spatial dimension of the grid
+ * \tparam TensorTags List of tags of the tensors to be monitored
+ */
+template <size_t Dim, typename TensorTags>
+class TruncationError : public Criterion {
+ public:
+  struct TargetTruncationError {
+    static std::string name() { return "Target"; }
+    using type = double;
+    static constexpr Options::String help = {
+        "The absolute target truncation error. If any tensor component "
+        "has a truncation error above this value, the element will be "
+        "p-refined."};
+  };
+
+  using options = tmpl::list<TargetTruncationError>;
+
+  static constexpr Options::String help = {
+      "Refine the grid towards the target truncation error"};
+
+  TruncationError() = default;
+
+  TruncationError(const double target_truncation_error)
+      : target_truncation_error_(target_truncation_error) {}
+
+  /// \cond
+  explicit TruncationError(CkMigrateMessage* msg) : Criterion(msg) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(TruncationError);  // NOLINT
+  /// \endcond
+
+  using compute_tags_for_observation_box = tmpl::list<>;
+
+  using argument_tags = tmpl::list<::Tags::DataBox>;
+
+  template <typename DbTagsList, typename Metavariables>
+  auto operator()(const db::DataBox<DbTagsList>& box,
+                  Parallel::GlobalCache<Metavariables>& /*cache*/,
+                  const ElementId<Dim>& /*element_id*/) const {
+    auto result = make_array<Dim>(Flag::DoNothing);
+    const auto& mesh = db::get<domain::Tags::Mesh<Dim>>(box);
+    tmpl::for_each<TensorTags>(
+        [&result, &box, &mesh,
+         target_truncation_error = target_truncation_error_](const auto tag_v) {
+          // Continue only if AMR flags are still unset
```

Comment:
Why? should there not be some sort of priority procedure over all the flags? If not, this should be documented.

Reply:
Added with the coarsening